lib/beaker/shared/host_manager.rb

```diff
@@ -100,8 +102,18 @@ def run_block_on hosts = [], filter = nil, &block
         end
         if block_hosts.is_a? Array
           if block_hosts.length > 0
-            result = block_hosts.map do |h|
-              run_block_on h, nil, &block
+            if opts[:run_in_parallel]
+              # Pass caller[1] - the line that called block_on - for logging purposes.
+              result = block_hosts.map.each_in_parallel(caller[1]) do |h|
+                run_block_on h, &block
+              end
+              hosts.each{|host| host.close}# For some reason, I have to close the SSH connection
+              # after spawning a process and running commands on a host,
+              # or else it gets into a broken state for the next call.
```

Comment:
This may have something to do with the behavior of the SSH library in use. Is it possible that the the SSH library closes the connections on process termination? In that case, the SSH library would probably only be deleting the fd from the child process' fd table--the original copy of the fd in the parent process would still be open even though the connection itself is closed. Closing the connection manually here would clean up that fd and allow the host object to properly reestablish the connection on the next attempt. For what it's worth, I am not personally familiar with the SSH connection code in use here, @KevinCorcoran or @tvpartytonight can provide more insight.